doc/ipv6/ipv6_link_local.md

```diff
@@ -0,0 +1,378 @@
+```
+
+### 3.2.5 COUNTER DB
+There are no changes to COUNTER DB schema definition.
+
+## 3.3 Switch State Service Design
+
+### 3.3.1 Orchestration Agent
+- Adds IP2ME /128 route for the interface link-local address to hardware (Only 1 IP2ME route is added as the link-local address is same on all IPv6 interfaces). This enables prioritizing the traffic destined to the interface link-local address.
+- Adds the fe80::/10 subnet route to the hardware per VRF. This allows for all traffic destined to link-local prefix to be copied to CPU.  Instead of adding multiple link-local routes to the hardware, a single /10 route is added to handle all link-local traffic. This helps reduce the usage of the hardware L3 route table resource.
+
+For all practical use-cases, the fe80::/10 route is enough to handle all IPv6 packets destined to link-local destination. The IP2ME/128 route is added to be consistent with global IPv6 address - as the underlying SAI implementation may treat these differently. Also, in some deployments, servers can be configured with a default gateway in the fe80:: range - the fe80::/10 route helps to route traffic from those servers.
+
+Changes done in NeighOrch:
+
+* Add the interface parameter to the next hop object key. This allows to add multiple IPv6 link local next hops with same IPv6 address. This change was originally implemented as part of the VRF implementation.
+
+```
+struct NextHopKey
+{
+    IpAddress           ip_address;     // neighbor IP address
+    string              alias;          // incoming interface alias
+}
+```
+
+```
+sonic# show ipv6 route
+
+S>*  2222::/64 [1/0] via fe80::5054:ff:fe03:6175, Ethernet0, 00:00:04
+  *                  via fe80::5054:ff:fe03:6175, Ethernet4, 00:00:04
+  *                  via fe80::5054:ff:fe03:6175, Ethernet8, 00:00:04
+
+127.0.0.1:6379> keys *NEIGH*
+ 1) "NEIGH_TABLE:Ethernet8:fe80::5054:ff:fe03:6175"
+ 2) "NEIGH_TABLE:Ethernet0:fe80::5054:ff:fe03:6175"
+ 3) "NEIGH_TABLE:Ethernet4:fe80::5054:ff:fe03:6175"
+
+```
+
+### 3.3.2 NeighSyncd
+
+Neighsyncd is updated to learn the IPv6 link-local neighbors. All link-local neighbors that are dynamically learned are programmed to ASIC_DB.
+
+### 3.3.3 IntfMgr
+In the absence of manually configured addresses on an interface, to enable/disable IPv6 mode, IntfMgr is modified to handle changes to the CONFIG DB `'ipv6_use_link_local_only'` setting. The IPv6 mode is enforced using the Linux system control utility to set the `'disable_ipv6'` parameter. The default value is also set so that new interfaces are created with default disable IPv6 setting.
+
+`sysctl -w net.ipv6.conf.default.disable_ipv6=1`
+
+Since the Linux kernel auto-generates the IPv6 link-local address per interface, netlink events for IPv6 address addition and deletion are handled by the IntfMgr. All netlink messages other than RTM_NEWADDR, RTM_DELADDR are ignored. It also ignores all addresses other than IPv6 link-local addresses.

```

Comment:
Currently `intfmgr`or other manager does not listen on events from kernel but handled by corresponding *syncd. Originally there was an `intfsyncd `but was later removed. We may want to revisit this section on how to handle netlink events for link-local address.

Reply:
We initially considered using intfsyncd. But that can result in multiple producers for the INTF:PREFIX table (intfmgrd for configured addresses and intfsyncd for learned addresses).
To avoid any race conditions and inconsistent behavior (like for ex., if VRF interface unbind when INTF and INTF:PREFIX tables are deleted by intfmgrd and if there are IPv6 LL address updates from intfsyncd at the same time), we enhanced the intfmgrd to push the netlink address updates into the APP_DB.

Reply:
I don't think we should re-introduce intfsyncd as before. However this approach of intfmgrd listening on netlink doesn't lgtm. Do you have any other proposal?

Reply:
@prsunny As long as we have an interface table e.g INTF_TABLE:<intf-name> in App-DB, creating a router interface in intfsorch for that interface in HW should be fine..right? why do we need to worry about handling multiple IPv6 LLAs with reference count.